lib/locale_threads.t

```diff
@@ -54,67 +470,1290 @@ SKIP: { # perl #127708
+                push @this_locales,
+                            $this_tests->{$category_name}{locale_name};
+            }
+        }
+
+        while (keys $this_tests->%*) {
+            foreach my $category_name (sort keys $this_tests->%*) {
+                my $this_category_tests = $this_tests->{$category_name};
+                my $test = shift
+                                $this_category_tests->{locale_tests}->@*;
+                print STDERR __FILE__, ': ', __LINE__, ': ', Dumper $test
+                                                                    if $debug;
+                if (! $test) {
+                    delete $this_tests->{$category_name};
+                    next;
+                }
+
+                $test->{category_name} = $category_name;
+                my $locale_name = $this_category_tests->{locale_name};
+                $test->{locale_name} = $locale_name;
+                $test->{codeset} =
+                                $locale_name_to_object{$locale_name}{codeset};
+
+                push @this_cooked_tests, $test;
+            }
+        }
+
+        push @cooked_tests, {
+                              thread => $i,
+                              categories => \@this_categories,
+                              locales => \@this_locales,
+                              tests => \@this_cooked_tests,
+                            };
+    }
+
+    my $all_tests_ref = \@cooked_tests;
+    my $all_tests_file = tempfile();
+
+    # Store the tests into a file, retrievable by the subprocess
+    use Storable;
+    if (! defined store($all_tests_ref, $all_tests_file)) {
+        die "Could not save the built-up data structure";
     }
 
-    # See if multiple threads can simultaneously change the locale, and give
-    # the expected radix results.  On systems without a comma radix locale,
-    # run this anyway skipping the use of that, to verify that we don't
-    # segfault
+    my $category_number_to_name = Data::Dumper->Dump(
+                                            [ \%map_category_number_to_name ],
+                                            [  'map_category_number_to_name']);
+
+    my $switches = "";
+    $switches = "switches => [ -DLv ]" if $debug > 2;
+
+    # Finally ready to run the test.  This stresses locale thread safety.  We
+    # start a bunch of threads.  Each sets the locale of each category being
+    # tested to the value determined in the code above.  Then each sleeps
+    # to a common start time, at which point they awaken and iterate their
+    # respective loops.  Each iteration runs a set of tests and checks that
+    # the results are as expected.  This should catch any instances of other
+    # threads interfering.  The code above has set up things so that each
+    # thread has as disparate results from the other threads as possible, so
+    # to more likely catch any bleed-through.
     fresh_perl_is("
+        BEGIN { \$| = 1; }
```

Comment:
Rather than a huge string literal here, I think it would be better to build this up in pieces, with heredocs for the literal parts.

Right now it's mostly literal text with many backslashes, making it hard to pick out the small number of replacements (like `$iterations_per_test_group`, `$category_number_to_name`).

Reply:
Done